### PR TITLE
Create cloudformation resources to support hit

### PR DESCRIPTION
**Pull request by demartinofra**
### Incorporate latest changes from custom-resource-helper

This simply imports the latest changes from `https://github.com/aws-cloudformation/custom-resource-helper`

### Create ResourcesS3Bucket when scheduler is Slurm

Enable the creation of a cluster dedicated S3 bucket when scheduler is `slurm`. This logic was already used to create an S3 bucket for the cluster to use when the scheduler is `awsbatch`

### Implement CustomResource to terminate ComputeFleet on stack deletion

We are going to have a single AWS Lambda function that contains the code to cleanup the s3 bucket and terminate the compute fleet. However we are going to have two different CFN CustomResources that can invoke such function in different moments by using different functionalities. 
A known issue with the CustomResource is that if it fails and returns an error to CloudFormation you cannot retry the execution (actually CloudFormation retries the execution once if I'm not wrong). Because of this we are going to add to the pcluster stop and delete some extra logic that also takes care of cleaning up instances in case they are still running.
The cleanup of the compute fleet must happen after the head node is terminated and before the placement groups are deleted otherwise we risk to violate constraints.

### Implement CloudFormation stack for heterogeneous instance types

This contains the definition of the CloudFormation template for HIT, implemented with Jinja directives

### Dynamically generate hit substack on pcluster create

This is a prototype of the code that dynamically generates the template at cluster creation time and uploads it to S3.
Here is a small sequence diagram that explains the flow:

ParallelCluster HIT - pcluster create (1)

### Rename hit-substack to compute-fleet-hit-substack

Just some renaming :)

### Update internal json confg representation for hit

Updating internal representation of cluster config

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/pcluster/update/update_command.py [hunk 1]
```diff
@@ -179,7 +179,7 @@ def _restore_cfn_only_params(cfn_boto3_client, args, cfn_params, stack_name, tar
 
     scheduler = cluster_section.get_param_value("scheduler")
     # Autofill DesiredSize cfn param
-    if not args.reset_desired:
+    if not args.reset_desired and scheduler != "slurm":
```
Comment (ddeidda) [inline cli/pcluster/update/update_command.py:182]: I see this condition on the slurm scheduler, meant to check if we are in a HIT cluster, is used in many places; I'm wondering this condition could be lost among all other scheduler checks which can be there for different reasons. What about being more explicit and adding a global function like ``is_hit_cluster()`` that can currently just do the same test on the scheduler but would isolate the rest of the code from this information (and will be also ready to work in the case other schedulers will be covered by HIT support)?

File: cli/pcluster/utils.py [hunk 2]
```diff
@@ -755,6 +755,9 @@ def get_base_additional_iam_policies():
 def get_cluster_capacity(stack_name):
     stack = get_stack(stack_name)
     scheduler = get_cfn_param(stack.get("Parameters", []), "Scheduler")
+    if scheduler == "slurm":
```
Comment (ddeidda) [inline cli/pcluster/utils.py:759]: Same as above.

File: cli/pcluster/update/update_command.py (hunk 1 above)
Comment (demartinofra) [reply cli/pcluster/update/update_command.py]: 👍🏻

File: cli/pcluster/update/update_command.py [hunk 3]
```diff
@@ -184,7 +184,7 @@ def _restore_cfn_only_params(cfn_boto3_client, args, cfn_params, stack_name, tar
     elif scheduler == "awsbatch":
         LOGGER.info("reset_desired flag does not work with awsbatch scheduler")
 
-    if scheduler == "awsbatch":
+    if scheduler in {"awsbatch", "slurm"}:
         # Autofill ResourcesS3Bucket cfn param
         params = utils.get_stack(stack_name, cfn_boto3_client).get("Parameters")
         cfn_params["ResourcesS3Bucket"] = utils.get_cfn_param(params, "ResourcesS3Bucket")
```
[the remark this answers is not in the document]
Comment (ddeidda) [reply cli/pcluster/update/update_command.py:190]: OK Let's talk about it.

[the remark this answers is not in the document]
Comment (ddeidda) [reply cli/pcluster/update/update_command.py:190]: The way the mappings file is currently used is more like a schema of the configuration. Let's talk about this to reach a quick agreement.

File: cli/pcluster/commands.py [hunk 4]
```diff
@@ -95,18 +95,19 @@ def create(args):  # noqa: C901 FIXME!!!
...
         stack_name = utils.get_stack_name(args.cluster_name)
 
-        # If scheduler is awsbatch or Slurm create bucket with resources
-        if cluster_section.get_param_value("scheduler") == "awsbatch":
-            batch_bucket_name = _create_bucket_with_resources(stack_name, pcluster_config.region, resources_dirs=["resources/batch", "resources/custom_resources"])
-            cfn_params["ResourcesS3Bucket"] = batch_bucket_name
-        elif cluster_section.get_param_value("scheduler") == "slurm":
-            batch_bucket_name = _create_bucket_with_resources(stack_name, pcluster_config.region, resources_dirs=["resources/custom_resources"])
-            cfn_params["ResourcesS3Bucket"] = batch_bucket_name
+        if cluster_section.get_param_value("scheduler") in ["awsbatch", "slurm"]:
```
[the remark this answers is not in the document]
Comment (ddeidda) [reply cli/pcluster/commands.py]: What about going one other step further and completely remove the scheduler check here? The ``_create_bucket_with_resources`` function could create the correct resources based on the current configuration and abstract the caller methods from this internal logic. Or let's talk about it if you think it's better to keep this external check.

[the remark this answers is not in the document]
Comment (demartinofra) [reply cli/pcluster/commands.py]: this was done already. resolving comment

File: cli/pcluster/update/update_command.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ddeidda) [reply cli/pcluster/update/update_command.py:190]: We reached the following agreement:
- add a new "visibility" attribute to parameters; automatic ones will have visibility = private
- in the from_file of the Param class, raise an exception if the parameter has been specified in the configuration file

File: cloudformation/aws-parallelcluster.cfn.json [hunk 5]
```diff
@@ -1502,6 +1544,62 @@
...
+            },
+            {
+              "Sid": "EC2",
+              "Effect": "Allow",
+              "Action": [
+                "ec2:DescribeInstances",
+                "ec2:DescribeLaunchTemplates",
+                "ec2:RunInstances",
+                "ec2:DescribeInstanceStatus",
+                "ec2:RunInstances"
```
Comment (rexcsn) [inline cloudformation/aws-parallelcluster.cfn.json:1576]: Missing `ec2:CreateTags` permission here, so `run-instances` call will fail. Also this `ec2:RunInstances` permission is redundant, please replace with `ec2:CreateTags`

**Pull request merged**